File: macos/Services/MCP/Tools/Tier3/ExecuteWriteQueryTool.swift
// ExecuteWriteQueryTool.swift
// Gridex
//
// MCP Tool: Execute arbitrary write SQL. Requires user approval.

import Foundation

struct ExecuteWriteQueryTool: MCPTool {
    let name = "execute_write_query"
    let description = "Execute one write SQL statement (INSERT/UPDATE/DELETE). Requires user approval. Only available in read-write mode. Multi-statement input is rejected."
    let tier = MCPPermissionTier.write

    let inputSchema: [String: Any] = [
        "type": "object",
        "properties": [
            "connection_id": [
                "type": "string",
                "description": "Connection identifier"
            ],
            "sql": [
                "type": "string",
                "description": "A single SQL statement to execute. Must not contain multiple statements."
            ],
            "params": [
                "type": "array",
                "description": "Parameters for placeholders"
            ]
        ],
        "required": ["connection_id", "sql"]
    ]

    func execute(params: JSONValue, context: MCPToolContext) async throws -> MCPToolResult {
        let connectionId = try extractConnectionId(from: params)

        guard let sql = params["sql"]?.stringValue else {
            throw MCPToolError.invalidParameters("sql is required")
        }

        // Syntactic checks run against a comment/literal-stripped copy so
        // payloads hidden in comments or string literals can't fool them.
        // The ORIGINAL sql is still what gets executed.
        let codeOnly = MCPSQLSanitizer.stripCommentsAndStrings(sql)
        let upperCode = codeOnly.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()

        let withoutTrailingSemi = upperCode.hasSuffix(";")
            ? String(upperCode.dropLast()).trimmingCharacters(in: .whitespacesAndNewlines)
            : upperCode
        if withoutTrailingSemi.contains(";") {
            throw MCPToolError.invalidParameters("Multiple statements are not allowed. Send one statement at a time.")
        }

        let permission = await context.checkPermission(tier: tier, connectionId: connectionId)
        if let error = permission.errorMessage {
            throw MCPToolError.permissionDenied(error)
        }

        if upperCode.hasPrefix("SELECT") || upperCode.hasPrefix("WITH") {
            throw MCPToolError.invalidParameters("Use the 'query' tool for SELECT / WITH statements. This tool is for write operations only.")
        }

        if upperCode.hasPrefix("UPDATE") || upperCode.hasPrefix("DELETE") {
            let range = NSRange(upperCode.startIndex..., in: upperCode)
            if Self.whereRegex.firstMatch(in: upperCode, options: [], range: range) == nil {
                throw MCPToolError.permissionDenied("UPDATE/DELETE without WHERE clause is not allowed.")
            }
            if let whereRange = codeOnly.range(of: "WHERE", options: .caseInsensitive) {
                let whereClause = String(codeOnly[whereRange.upperBound...])
                    .trimmingCharacters(in: .whitespaces)
                if let error = await context.permissionEngine.validateWhereClause(whereClause).errorMessage {
                    throw MCPToolError.permissionDenied(error)
                }
            }
        }

        let (adapter, _) = try await context.getAdapter(for: connectionId)

        if permission.requiresUserApproval {
            let approved = await context.requestApproval(
                tool: name,
                description: "Execute write query",
                details: "SQL:\n\(sql)",
                connectionId: connectionId
            )
            if !approved {
                throw MCPToolError.permissionDenied("User denied the operation.")
            }
        }

        let queryParams = params["params"]?.arrayValue ?? []
        let result: QueryResult
        if queryParams.isEmpty {
            result = try await adapter.executeRaw(sql: sql)
        } else {
            let rowParams = queryParams.map { jsonValueToRowValue($0) }
            result = try await adapter.executeWithRowValues(sql: sql, parameters: rowParams)
        }

        var response = "Query executed successfully."
        if result.rowsAffected > 0 {
            response += " \(result.rowsAffected) row(s) affected."
        }
        return MCPToolResult(text: response)
    }

    private static let whereRegex = try! NSRegularExpression(pattern: "\\bWHERE\\b", options: .caseInsensitive)

    private func jsonValueToRowValue(_ value: JSONValue) -> RowValue {
        switch value {
        case .null: return .null
        case .bool(let b): return .boolean(b)
        case .int(let i): return .integer(Int64(i))
        case .double(let d): return .double(d)
        case .string(let s): return .string(s)
        case .array, .object:
            let encoder = JSONEncoder()
            if let data = try? encoder.encode(value),
               let str = String(data: data, encoding: .utf8) {
                return .json(str)
            }
            return .null
        }
    }
}

File: windows/Gridex/Services/MCP/Tools/Tier3/ExecuteWriteQueryTool.cpp
//
// ExecuteWriteQueryTool.cpp
//

#include "ExecuteWriteQueryTool.h"
#include "../MCPToolHelpers.h"
#include "../../Security/MCPSQLSanitizer.h"
#include <algorithm>
#include <regex>

namespace DBModels
{
    namespace
    {
        std::string trim(const std::string& s)
        {
            size_t a = 0, b = s.size();
            while (a < b && std::isspace(static_cast<unsigned char>(s[a]))) ++a;
            while (b > a && std::isspace(static_cast<unsigned char>(s[b - 1]))) --b;
            return s.substr(a, b - a);
        }
        std::string upper(std::string s)
        {
            std::transform(s.begin(), s.end(), s.begin(),
                [](unsigned char c){ return static_cast<char>(std::toupper(c)); });
            return s;
        }
        bool startsWith(const std::string& s, const std::string& prefix)
        {
            return s.size() >= prefix.size() &&
                   std::equal(prefix.begin(), prefix.end(), s.begin());
        }
    }

    MCPToolResult ExecuteWriteQueryTool::execute(const nlohmann::json& params, MCPToolContext& ctx)
    {
        const auto connId = MCPTool::extractConnectionId(params);

        if (!params.contains("sql") || !params["sql"].is_string())
            throw MCPToolError::invalidParameters("sql is required");
        const auto sqlUtf8 = params["sql"].get<std::string>();

        // Sanitize + multi-statement guard. Run syntactic checks
        // against a comment/literal-stripped copy so payloads
        // hidden inside comments or strings cannot fool them.
        const std::string codeOnly = MCPSQLSanitizer::stripCommentsAndStrings(sqlUtf8);
        const std::string trimmedUpper = upper(trim(codeOnly));

        std::string withoutTrailingSemi = trimmedUpper;
        if (!withoutTrailingSemi.empty() && withoutTrailingSemi.back() == ';')
        {
            withoutTrailingSemi.pop_back();
            withoutTrailingSemi = trim(withoutTrailingSemi);
        }
        if (withoutTrailingSemi.find(';') != std::string::npos)
            throw MCPToolError::invalidParameters(
                "Multiple statements are not allowed. Send one statement at a time.");

        if (startsWith(trimmedUpper, "SELECT") || startsWith(trimmedUpper, "WITH"))
            throw MCPToolError::invalidParameters(
                "Use the 'query' tool for SELECT / WITH statements. "
                "execute_write_query is for write operations only.");

        // UPDATE/DELETE must contain WHERE and the WHERE clause
        // must pass the permission-engine guard (no `;`, no
        // comments, no trivial `1=1`).
        if (startsWith(trimmedUpper, "UPDATE") || startsWith(trimmedUpper, "DELETE"))
        {
            static const std::regex whereRe(R"(\bWHERE\b)", std::regex::icase);
            if (!std::regex_search(trimmedUpper, whereRe))
                throw MCPToolError::permissionDenied(
                    "UPDATE/DELETE without WHERE clause is not allowed.");

            // Extract the WHERE tail from the sanitized SQL to feed
            // the WHERE validator. Case-insensitive search in the
            // sanitized text, slice everything after it.
            std::smatch m;
            if (std::regex_search(codeOnly, m, whereRe))
            {
                const auto whereTail = trim(std::string(
                    m.suffix().first, m.suffix().second));
                const auto v = ctx.permissionEngine.validateWhereClause(whereTail);
                if (const auto* msg = v.errorMessage())
                    throw MCPToolError::permissionDenied(*msg);
            }
        }

        const auto perm = ctx.checkPermission(tier(), connId);
        if (const auto* msg = perm.errorMessage())
            throw MCPToolError::permissionDenied(*msg);

        auto [adapter, _] = ctx.getAdapter(connId);

        if (perm.requiresUserApproval())
        {
            const auto clientW = MCPToolHelpers::fromUtf8(
                ctx.client.name.empty() ? std::string{"AI client"} : ctx.client.name);
            const std::wstring details = L"SQL:\n" + MCPToolHelpers::fromUtf8(sqlUtf8);
            const bool ok = ctx.requestApproval(
                MCPToolHelpers::fromUtf8(this->name()),
                L"Execute write query",
                details, connId, clientW);
            if (!ok) throw MCPToolError::permissionDenied("User denied the operation.");
        }

        QueryResult r;
        try { r = adapter->execute(MCPToolHelpers::fromUtf8(sqlUtf8)); }
        catch (const std::exception& e)
        { throw MCPToolError::queryFailed(e.what()); }
        if (!r.success) throw MCPToolError::queryFailed(MCPToolHelpers::toUtf8(r.error));

        std::string resp = "Query executed successfully.";
        if (r.totalRows > 0)
            resp += " " + std::to_string(r.totalRows) + " row(s) affected.";
        return MCPToolResult::text(resp);
    }
}
